occulow-firmware/src/drivers/people_counting/people_counting.c

```diff
@@ -0,0 +1,215 @@
+*
+* Created: 3/24/2017 10:51:44 PM
+*  Author: Udaya Malik
+*/
+
+#include <asf.h>
+#include <drivers/grideye/grideye.h>
+#include <drivers/people_counting/people_counting_utils.h>
+#include <drivers/people_counting/people_counting_internal.h>
+
+// Allocate areas to store raw and median frames
+static frame_elem_t MEDIAN_FRAME_CHUNK[MEDIAN_FRAME_CHUNK_SIZE];
+static frame_elem_t RAW_FRAME_CHUNK[RAW_FRAME_CHUNK_SIZE];
+
+// Allocate lists of pointers into the chunks
+static frame_t RAW_FRAMES[NUM_RAW_FRAMES];
+static frame_t MEDIAN_FRAMES[NUM_MEDIAN_FRAMES];
+
+static uint16_t raw_frame_count = 0;  ///< Number of frames in the RAW_FRAMES
+static bool raw_frames_filled = false;  ///< Whether or not RAW_FRAMES is full
+static pc_counter_t pc_counter;  ///< Counter object
+static pc_config_t pc_config;  ///< Configuration object
+
+
+/**
+ * @brief      Initializes the people counting driver
+ */
+void pc_init() {
+	initialize_counter(&pc_counter);
+	initialize_config(&pc_config);
+	initialize_frame_stacks();
+}
+
+/**
+ * @brief      Adds a new frame to be processed (for people counting)
+ *
+ * @param[in]  new_frame  The new frame to be processed
+ */
+void pc_new_frame(frame_t new_frame) {
+	// Check if RAW_FRAMES has been filled
+	if (raw_frame_count < NUM_RAW_FRAMES) {
+		raw_frame_count++;
+	} else {
+		raw_frames_filled = true;
+	}
+	
+	// Enqueue new_frame into RAW_FRAMES
+	enqueue_frame(RAW_FRAMES, new_frame, NUM_RAW_FRAMES);
+	
+	// Compute and enqueue median frame
+	frame_elem_t median_filtered_frame[GE_FRAME_SIZE];
+	compute_median_frame(median_filtered_frame, RAW_FRAMES, NUM_RAW_FRAMES);
+	
+	// Subtract median from the new frame
+	for (int i = 0; i < GE_FRAME_SIZE; i++) {
+		if (new_frame[i] < median_filtered_frame[i]) {
+			median_filtered_frame[i] = 0;
+		} else {
+			median_filtered_frame[i] = new_frame[i] - median_filtered_frame[i];
+		}
+	}
+	
+	enqueue_frame(MEDIAN_FRAMES, median_filtered_frame, NUM_MEDIAN_FRAMES);
+	
+	// Reset counted
+	pc_counter.count_updated = false;
+}
+
+/**
+ * @brief      Gets the ingress count
+ *
+ * @return     The ingress count
+ */
+double pc_get_in_count(void) {
+	// Update count if not updated
+	if (!pc_counter.count_updated) {
+		update_counter();
+	}
+	return pc_counter.in_count;
+}
+
+/**
+ * @brief      Gets the egress count
+ *
+ * @return     The egress count
+ */
+double pc_get_out_count(void) {
+	// Update count if not updated
+	if (!pc_counter.count_updated) {
+		update_counter();
+	}
+	return pc_counter.out_count;
+}
+
+/**
+ * @brief      Initializes a counter object
+ */
+static void initialize_counter(pc_counter_t *counter)
+{
+	counter->in_count = 0;
+	counter->out_count = 0;
+	counter->count_updated = false;
+}
+
+static void initialize_config(pc_config_t *config) {
+	config->trigger_column[0] = TRIGGER_COLUNM_2;
+	config->trigger_column[1] = TRIGGER_COLUNM_5;
+	config->trigger_check_offset[0] = CHECK_OFFSET_2;
+	config->trigger_check_offset[1] = CHECK_OFFSET_5;

```

Comment:
I changed `CHECK_OFFSET_` to be more descriptive. I've left the trigger columns as is, since I think the macros `TRIGGER_COLUMN_0` thru `TRIGGER_COLUMN_7` could technically be defined and accurately describe what they do.